docs: snapshots, binding context, yaml config, queues, debug commands (#118)
flant/shell-operator: HOOKS.md

- If hook's configuration is successful, the working queue named "main" is filled with `onStartup` hooks.

- Then, the "main" queue is filled with `kubernetes` hooks with `Synchronization` [binding context](#binding-context) type, so that each hook receives all existing objects described in hook's configuration.

- After executing `kubernetes` hook with `Synchronization` binding context, Shell-operator starts a monitor of Kubernetes events according to configured `kubernetes` binding.
  - Each monitor stores a *snapshot* — a refreshable list of all Kubernetes objects that match a binding definition.

Next the main cycle is started:

- Event handler adds hooks to the named queues on events:
  - `kubernetes` hooks are added to the queue when desired [WatchEvent](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#watchevent-v1-meta) is received from Kubernetes,
  - `schedule` hooks are added according to the schedule,
  - `kubernetes` and `schedule` hooks are added to the "main" queue or to the named queue if `queue` field was specified.

> - Each named queue has an own queue handler which executes hooks strictly sequentially. If hook fails with an error (non-zero exit code), Shell-operator restarts it (every 5 seconds) until success. In case of an erroneous execution of some hook, when other events occur, the queue will be filled with new tasks, but their execution will be blocked until the failing hook succeeds.

Reviewer: ```suggestion
- Each named queue has its own queue handler which executes hooks strictly sequentially. If hook fails with an error (non-zero exit code), Shell-operator restarts it (every 5 seconds) until succeeds. In case of an erroneous execution of some hook, when other events occur, the queue will be filled with new tasks, but their execution will be blocked until the failing hook succeeds.
```

  - You can change this behavior for a specific hook by adding `allowFailure: true` to the binding configuration (not available for `onStartup` hooks).
  

> - Each hook is executed with binding context, that describe an occurred event: 

Reviewer: ```suggestion
- Each hook is executed with a binding context, that describes an already occurred event: 
```

- `namespace.labelSelector` — this filter works like `labelSelector` but for namespaces and Shell-operator dynamically subscribes to events from matched namespaces.

> - `jqFilter` —  an optional parameter that specifies *event filtering* with [jq syntax](https://stedolan.github.io/jq/manual/). The hook will be triggered on Modified event only if the filter result is changed after the last event. See example [102-monitor-namespaces](examples/102-monitor-namespaces).

Reviewer: ```suggestion
- `jqFilter` —  an optional parameter that specifies *event filtering* using [jq syntax](https://stedolan.github.io/jq/manual/). The hook will be triggered on Modified event only if the filter result is changed after the last event. See example [102-monitor-namespaces](examples/102-monitor-namespaces).
```
